onnxruntime/core/providers/webnn/builders/impl/attention_op_builder.cc

```diff
@@ -0,0 +1,568 @@
+  emscripten::val value_input = model_builder.GetOperand(input_defs[2]->Name());
+  emscripten::val past_key_input = model_builder.GetOperand(input_defs[3]->Name());
+  emscripten::val past_value_input = model_builder.GetOperand(input_defs[4]->Name());
+  emscripten::val seqlens_k_input = model_builder.GetOperand(input_defs[5]->Name());
+
+  std::vector<int64_t> input_q_shape, input_k_shape, input_v_shape, input_past_k_shape, input_past_v_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_q_shape, logger), "Cannot get query shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[1], input_k_shape, logger), "Cannot get key shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[2], input_v_shape, logger), "Cannot get value shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[3], input_past_k_shape, logger), "Cannot get past_key shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[4], input_past_v_shape, logger), "Cannot get past_value shape");
+  const auto q_rank = input_q_shape.size();
+  const auto k_rank = input_k_shape.size();
+  const auto v_rank = input_v_shape.size();
+  const auto past_k_rank = input_past_k_shape.size();
+  const auto past_v_rank = input_past_v_shape.size();
+  ORT_RETURN_IF_NOT(q_rank == 3 && k_rank == 3 && v_rank == 3,
+                    "The qkv shape should be batch_size * sequence_length * hidden_size.");
+  ORT_RETURN_IF_NOT(past_k_rank == 4 && past_v_rank == 4, "The qkv shape should be BNSH.");
+
+  NodeAttrHelper helper(node);
+  uint32_t kv_num_heads = helper.Get("kv_num_heads", 0);
+  uint32_t num_heads = helper.Get("num_heads", 0);
+  ORT_RETURN_IF_NOT(num_heads != 0 && kv_num_heads != 0, "Attributes num_heads and kv_num_heads are required.");
+
+  uint32_t batch_size = SafeInt<uint32_t>(input_q_shape[0]);
+  uint32_t qkv_sequence_length = SafeInt<uint32_t>(input_q_shape[1]);
+  uint32_t qkv_hidden_size = SafeInt<uint32_t>(input_q_shape[2]);
+  uint32_t head_size = SafeInt<uint32_t>(qkv_hidden_size / num_heads);
+  uint32_t past_sequence_length = SafeInt<uint32_t>(input_past_k_shape[2]);
+  uint32_t group_size = SafeInt<uint32_t>(num_heads / kv_num_heads);
+
+  std::vector<uint32_t> reshape_output_shape = {batch_size, qkv_sequence_length, qkv_hidden_size};
+  std::vector<uint32_t> scatter_indices_shape = {batch_size, qkv_sequence_length, kv_num_heads, 3};
+  std::vector<uint32_t> reshape_tensor_shape = {batch_size, qkv_sequence_length, num_heads, head_size};
+
+  emscripten::val common_options = emscripten::val::object();
+  if (input_defs[0]->Type() == onnx::Utils::DataTypeUtils::ToType("float16")) {
+    common_options.set("label", node.Name() + "/GQA/preprocess/cast/query_input");
+    query_input = model_builder.GetBuilder().call<emscripten::val>("cast", query_input, emscripten::val("float32"),
+                                                                   common_options);
+  }
+  if (input_defs[1]->Type() == onnx::Utils::DataTypeUtils::ToType("float16")) {
```

Comment:
Duplicated `float16` data type check, as we have make sure all the inputs same in `HasSupportedInputsImpl`.